Add debuggable version of the core

xobs opened the pull request:
I'm working with the NeTV2 project, and have discovered that it's really handy to have a version of the VexRiscv core around with the debug plugin enabled.

This checks in a separate version of the core that is identical, except the DebugPlugin is uncommented.

Dolu1990 commented:
I agree it is usefull, then i'm not sure what is the best way to have multiple core versions.

Maybe we can add programm arguments to say if we want debug or not ? cache size, and other options ?

ex : sbt compile "run-main vexriscv.GenCoreDefault --debug --i$size=4096" ?

sbourdeauducq commented:
> This checks in a separate version of the core that is identical, except the DebugPlugin is uncommented.

But this removes the code that was used to generate the original version. Can you solve this issue?

xobs commented:
@sbourdeauducq I've checked in a separate GenCoreDefaultDebug.scala that simply adds the Debug plugin.

@Dolu1990 Do you have an example of how this check would be done at build time?

Dolu1990 commented:
@xobs 

There is a generic example : 

```scala
case class ArgConfig(debug : Boolean = false,
                     iCacheSize : Int = 4096)

object PlayDevConsole{
  def main(args: Array[String]): Unit = {
    val parser = new scopt.OptionParser[ArgConfig]("VexRiscvGen") {
      //  ex :-d    or   --debug
      opt[Unit]('d', "debug")    action { (_, c) => c.copy(debug = true)   } text("Enable debug")
      // ex : -iCacheSize=XXX
      opt[Int]("iCacheSize")     action { (v, c) => c.copy(iCacheSize = v) } text("Set instruction cache size")
    }

    val argConfig = parser.parse(args, ArgConfig()).get

    println(argConfig)
  }
}
```

And there is an example to configure VexRiscv with it (it had some move around to make the list of plugins programaticaly accesible : 

```scala
package vexriscv

import spinal.core._
import spinal.lib._
import vexriscv.ip.{DataCacheConfig, InstructionCacheConfig}
import vexriscv.plugin.CsrAccess.WRITE_ONLY
import vexriscv.plugin._

import scala.collection.mutable.ArrayBuffer

object SpinalConfig extends spinal.core.SpinalConfig(
  defaultConfigForClockDomains = ClockDomainConfig(
    resetKind = spinal.core.SYNC
  )
)

case class ArgConfig(
  debug : Boolean = false,
  iCacheSize : Int = 4096
)

object GenCoreDefault{
  def main(args: Array[String]) {
    SpinalConfig.generateVerilog{
      //Parse arguments
      val parser = new scopt.OptionParser[ArgConfig]("VexRiscvGen") {
        //  ex :-d    or   --debug
        opt[Unit]('d', "debug")    action { (_, c) => c.copy(debug = true)   } text("Enable debug")
        // ex : -iCacheSize=XXX
        opt[Int]("iCacheSize")     action { (v, c) => c.copy(iCacheSize = v) } text("Set instruction cache size")
      }
      val argConfig = parser.parse(args, ArgConfig()).get

      //Gen CPU plugin list
      val plugins = ArrayBuffer[Plugin[VexRiscv]]()
      plugins ++= List(
        new PcManagerSimplePlugin(
          resetVector = null, //null => external
          relaxedPcCalculation = false
        ),
        new IBusCachedPlugin(
          config = InstructionCacheConfig(
            cacheSize = argConfig.iCacheSize,
            bytePerLine =32,
            wayCount = 1,
            addressWidth = 32,
            cpuDataWidth = 32,
            memDataWidth = 32,
            catchIllegalAccess = true,
            catchAccessFault = true,
            catchMemoryTranslationMiss = true,
            asyncTagMemory = false,
            twoCycleRam = true
          )
        ),
        new DBusCachedPlugin(
          config = new DataCacheConfig(
            cacheSize         = 4096,
            bytePerLine       = 32,
            wayCount          = 1,
            addressWidth      = 32,
            cpuDataWidth      = 32,
            memDataWidth      = 32,
            catchAccessError  = true,
            catchIllegal      = true,
            catchUnaligned    = true,
            catchMemoryTranslationMiss = true
          ),
          memoryTranslatorPortConfig = null,
          csrInfo = true
        ),
        new StaticMemoryTranslatorPlugin(
          ioRange      = _.msb
        ),
        new DecoderSimplePlugin(
          catchIllegalInstruction = true
        ),
        new RegFilePlugin(
          regFileReadyKind = plugin.SYNC,
          zeroBoot = false
        ),
        new IntAluPlugin,
        new SrcPlugin(
          separatedAddSub = false,
          executeInsertion = true
        ),
        new FullBarrielShifterPlugin,
        new MulPlugin,
        new DivPlugin,
        new HazardSimplePlugin(
          bypassExecute           = true,
          bypassMemory            = true,
          bypassWriteBack         = true,
          bypassWriteBackBuffer   = true,
          pessimisticUseSrc       = false,
          pessimisticWriteRegFile = false,
          pessimisticAddressMatch = false
        ),
        new BranchPlugin(
          earlyBranch = false,
          catchAddressMisaligned = true,
          prediction = STATIC
        ),
        new CsrPlugin(
          config = CsrPluginConfig.small(mtvecInit = null).copy(mtvecAccess = WRITE_ONLY)
        ),
        new ExternalInterruptArrayPlugin(),
        new YamlPlugin("cpu0.yaml")
      )

      if(argConfig.debug){
        plugins += new DebugPlugin(ClockDomain.current.clone(reset = Bool().setName("debugReset")))
      }

      //CPU configuration
      val cpuConfig = VexRiscvConfig(plugins.toList)

      //CPU instanciation
      val cpu = new VexRiscv(cpuConfig)

      //CPU modifications to be an Wishbone one
      cpu.rework {
        for (plugin <- cpuConfig.plugins) plugin match {

          case plugin: IBusCachedPlugin => {
            plugin.iBus.asDirectionLess() //Unset IO properties of iBus
            master(plugin.iBus.toWishbone()).setName("iBusWishbone")
          }
          case plugin: DBusCachedPlugin => {
            plugin.dBus.asDirectionLess()
            master(plugin.dBus.toWishbone()).setName("dBusWishbone")
          }
          case _ =>
        }
      }
      cpu
    }
  }
}
```

Which can then be used this way (for instance) : 

```
sbt compile "run-main vexriscv.GenCoreDefault -d --iCacheSize=1024"
```

xobs commented:
@Dolu1990 I've updated GenCoreDefault.scala with parameters for debug mode, icachesize, and dcachesize.  Additionally, I've updated the README to show examples of how to use it.

Dolu1990 commented:
Thanks ^^
